src/ess/estia/calibration.py

```diff
@@ -0,0 +1,239 @@
+import numpy as np
+import scipp as sc
+
+from ..reflectometry.normalization import (
+    reduce_from_events_to_lz,
+    reduce_to_q,
+)
+from ..reflectometry.types import CoordTransformationGraph, QBins, WavelengthBins
+from .types import (
+    Intensity,
+    MagneticReference,
+    MagneticSample,
+    NonMagneticReference,
+    OffOff,
+    OffOn,
+    OnOff,
+    OnOn,
+    PolarizedReflectivityOverQ,
+)
+
+
+def solve_for_calibration_parameters(Io, Is):
+    """
+    Solves for the calibration parameters given the reference
+    measurements.
+
+    See https://doi.org/10.1016/S0921-4526(00)00823-1.
+    """
+    Iopp, Iopa, Ioap, Ioaa = Io
+    Ipp, Ipa, Iap, Iaa = Is
+
+    I0 = 2 * (Iopp * Ioaa - Iopa * Ioap) / (Iopp + Ioaa - Iopa - Ioap)
+    rho = (Ioaa - Ioap) / (Iopp - Iopa)
+    alp = (Ioaa - Iopa) / (Iopp - Ioap)
+
+    Rspp_plus_Rsaa = (
+        4
+        * (rho * alp * Ipp + Iaa + rho * Ipa + alp * Iap)
+        / ((1 + rho) * (1 + alp) * I0)
+    )
+    Pp = sc.sqrt(
+        (Ipp + Iaa - Ipa - Iap)
+        * (alp * (Ipp - Iap) - Iaa + Ipa)
+        / (
+            (rho * alp * Ipp + Iaa + rho * Ipa + alp * Iap)
+            * (rho * (Ipp - Ipa) - Iaa + Iap)
+        )
+    )
+    Ap = sc.sqrt(
+        (Ipp + Iaa - Ipa - Iap)
+        * (rho * (Ipp - Ipa) - Iaa + Iap)
+        / (
+            (rho * alp * Ipp + Iaa + rho * Ipa + alp * Iap)
+            * (alp * (Ipp - Iap) - Iaa + Ipa)
+        )
+    )
+    Rs = sc.sqrt(
+        (alp * (Ipp - Iap) - Iaa + Ipa)
+        * (rho * (Ipp - Ipa) - Iaa + Iap)
+        / ((rho * alp * Ipp + Iaa + rho * Ipa + alp * Iap) * (Ipp + Iaa - Ipa - Iap))
+    )
+
+    Pa = -rho * Pp
+    Aa = -alp * Ap
+
+    Rspp_minus_Rsaa = Rs * Rspp_plus_Rsaa
+    Rspp = (Rspp_plus_Rsaa + Rspp_minus_Rsaa) / 2
+    Rsaa = Rspp_plus_Rsaa - Rspp
+
+    return I0, Pp, Pa, Ap, Aa, Rspp, Rsaa
+
+
+def correction_matrix(Pp, Pa, Ap, Aa):
+    """
+    Defines the linear relationship between measured intensity
+    and reflectivity.
+    """
+    return [
+        [
+            (1 + Pp) * (1 + Ap) / 4,
+            (1 + Pp) * (1 - Ap) / 4,
+            (1 - Pp) * (1 + Ap) / 4,
+            (1 - Pp) * (1 - Ap) / 4,
+        ],
+        [
+            (1 + Pp) * (1 + Aa) / 4,
+            (1 + Pp) * (1 - Aa) / 4,
+            (1 - Pp) * (1 + Aa) / 4,
+            (1 - Pp) * (1 - Aa) / 4,
+        ],
+        [
+            (1 + Pa) * (1 + Ap) / 4,
+            (1 + Pa) * (1 - Ap) / 4,
+            (1 - Pa) * (1 + Ap) / 4,
+            (1 - Pa) * (1 - Ap) / 4,
+        ],
+        [
+            (1 + Pa) * (1 + Aa) / 4,
+            (1 + Pa) * (1 - Aa) / 4,
+            (1 - Pa) * (1 + Aa) / 4,
+            (1 - Pa) * (1 - Aa) / 4,
+        ],
+    ]
```

Comment:
Would this be more readable if we define two `2x2` matrices and use something like https://numpy.org/doc/stable/reference/generated/numpy.kron.html?